CountMinSketch/CountMinSketch.playground/Contents.swift

```diff
@@ -0,0 +1,111 @@
+//: # CountMin Sketch
+import Foundation
+
+/// Private wrapper around Hashing, allowing hash different Hashables and keep their value
+private final class Hashing<T> where T: Hashable {
+    private var map: [T: Int] = [:]
+    
+    func hash(_ value: T) -> Int {
+        if let hash = map[value] {
+            return hash
+        }
+        var hasher = Hasher()
+        hasher.combine(value)
+        let newValue = abs(hasher.finalize())
+        map[value] = newValue
+        return newValue
+    }
+}
+
+/*
+ A class for counting hashable items using the Count-min Sketch strategy.
+ It fulfills a similar purpose than `itertools.Counter`.
+ The Count-min Sketch is a randomized data structure that uses a constant
+ amount of memory and has constant insertion and lookup times at the cost
+ of an arbitrarily small overestimation of the counts.
+*/
+public final class CountMinSketch<T> where T: Hashable {
+    private var hashers: [Hashing<T>] = []
+    private var matrix: [[UInt64]] = []
+    private let rows: Int
+    private let cols: Int
+    
+    /// The total amount of elements adedd to the model
+    private(set) var count: UInt64 = 0
+    /// init - will determine the matrix size
+    /// - Parameters:
+    ///   - rows: the size of the hash tables, larger implies smaller overestimation
+    ///   - cols: the amount of hash tables, larger implies lower probability of
+    init(rows: Int, cols: Int) {
+        self.rows = rows
+        self.cols = cols
+        for _ in 0..<self.rows {
+            hashers.append(Hashing())
+            matrix.append([UInt64](repeating: 0, count: self.cols))
+        }
+    }
+    
+    /// Init - will determine the matrix size. s.t CountMin sketch guarantees approximation error on point queries more than epsilon * F1 (where F1 is the Frequency of first order of the stream) with probability `delta` in space O(1 \ epsilon * log(1 \ delta))
+    /// - Parameters:
+    ///   - delta: the probability for an error bigger than epsilon
+    ///   - epsilon: the error from the actual value
+    init(delta: CGFloat, epsilon: CGFloat) {
+        self.rows = Int(log2(1/delta).rounded(.up))
+        self.cols = Int((2/epsilon).rounded(.up))
+        for _ in 0..<self.rows {
+            hashers.append(Hashing())
+            matrix.append([UInt64](repeating: 0, count: self.cols))
+        }
+    }
+    
+    // Adding elemets to count, by default we increase the element count by one
+    // But we extended the API to allow increasing the count in batches
+    
+    /// Adding an element ot the sketch
```

Comment:
minor: typo - **ot** . It should be **to**

Reply:
Sorry for the delay - fixed, thanks!